From 3d1fe918585162d813727d0bccd9485e47b87aaa Mon Sep 17 00:00:00 2001
From: Vadim Rutkovsky <vrutkovs@redhat.com>
Date: Tue, 14 Jun 2016 17:37:29 +0200
Subject: [PATCH] Add install target

---
 GNUmakefile | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/GNUmakefile b/GNUmakefile
index 38656c8..6f6c941 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -11,4 +11,11 @@ build:
 clean:
 	./mach clobber
 
-.PHONY: all build clean
+install:
+	$(MAKE) -f client.mk install $@
+	desktop-file-install --dir $(prefix)/share/applications org.mozilla.Firefox.desktop
+	mkdir -p $(prefix)/share/icons/hicolor/128x128/apps/ $(prefix)/share/appdata
+	#cp -p browser/branding/aurora/default16.png $(prefix)/share/icons/hicolor/16x16/apps/org.mozilla.Firefox.png
+	cp -p org.mozilla.Firefox.appdata.xml $(prefix)/share/appdata
+	cp -p browser/branding/nightly/mozicon128.png $(prefix)/share/icons/hicolor/128x128/apps/org.mozilla.Firefox.png
+.PHONY: all build clean install
-- 
2.5.5

