From 3d1fe918585162d813727d0bccd9485e47b87aaa Mon Sep 17 00:00:00 2001
From: Vadim Rutkovsky <vrutkovs@redhat.com>
Date: Tue, 14 Jun 2016 17:37:29 +0200
Subject: [PATCH] Add install target

---
 GNUmakefile | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/GNUmakefile b/GNUmakefile
index 38656c8..6f6c941 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -11,4 +11,8 @@ build:
 clean:
 	./mach clobber
 
-.PHONY: all build clean
+install:
+	$(MAKE) -f client.mk install $@
+	desktop-file-install --dir $(prefix)/share/applications firefox.desktop
+
+.PHONY: all build clean install
-- 
2.5.5

