diff -up rustc-2/configure.orig rustc-2/configure
--- rustc-2/configure.orig	2017-01-25 10:19:24.946345433 +0100
+++ rustc-2/configure	2017-01-25 10:19:46.233512769 +0100
@@ -0,0 +1,2 @@
+# empty configure
+exit 0
diff -up rustc-2/Makefile.orig rustc-2/Makefile
--- rustc-2/Makefile.orig	2017-01-25 10:19:19.377301659 +0100
+++ rustc-2/Makefile	2017-01-25 10:15:42.225594669 +0100
@@ -0,0 +1,4 @@
+all:
+	echo "None"
+install:
+	./install.sh --prefix=/app
