diff -up firefox-4/testing/mozbase/mozlog/setup.py.blessings firefox-4/testing/mozbase/mozlog/setup.py
--- firefox-4/testing/mozbase/mozlog/setup.py.blessings	2017-03-03 10:02:50.826586940 +0100
+++ firefox-4/testing/mozbase/mozlog/setup.py	2017-03-03 10:02:56.062627745 +0100
@@ -17,7 +17,7 @@ setup(name=PACKAGE_NAME,
       license='MPL 1.1/GPL 2.0/LGPL 2.1',
       packages=find_packages(),
       zip_safe=False,
-      install_requires=["blessings>=1.3"],
+      #install_requires=["blessings>=1.3"],
       tests_require=['mozfile'],
       platforms=['Any'],
       classifiers=['Development Status :: 4 - Beta',
